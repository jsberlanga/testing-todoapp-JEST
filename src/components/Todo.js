import React, { useState } from "react";
import PropTypes from "prop-types";

const Todo = ({ todo, handleComplete, handleDelete, handleUpdate }) => {
  const [isEdit, setIsEdit] = useState(false);
  const [title, setTitle] = useState("");
  const onUpdate = ({ title }) => {
    setIsEdit(true);
    setTitle(title);
  };

  return (
    <div className="todo">
      {isEdit && (
        <div className="todo--option">
          <input
            type="text"
            value={title}
            onChange={e => setTitle(e.target.value)}
          />
          <button
            type="submit"
            onClick={() => {
              handleUpdate(todo, title);
              setIsEdit(false);
            }}
          >
            Update
          </button>
        </div>
      )}
      {!isEdit && (
        <div className="todo--option">
          <button data-test="edit-todo" onClick={() => onUpdate(todo)}>
            edit
          </button>
          <p
            data-test="complete-todo"
            onClick={() => handleComplete(todo)}
            className={todo.complete ? "todo__complete" : ""}
          >
            {todo.title}
          </p>
          <button data-test="delete-todo" onClick={() => handleDelete(todo)}>
            x
          </button>
        </div>
      )}
    </div>
  );
};

Todo.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    complete: PropTypes.bool.isRequired
  }).isRequired
};

export default Todo;
