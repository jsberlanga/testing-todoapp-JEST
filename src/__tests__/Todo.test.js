import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Todo from "../components/Todo";

afterEach(cleanup);

const fakeTodo = {
  id: 1,
  title: "fake todo",
  complete: true
};

// afterEach(cleanup);

describe("<Todo />", () => {
  test("renders todo", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Todo todo={fakeTodo} />, div);
    expect(div).toMatchSnapshot();
  });
  test("changes class when complete", () => {
    const component = renderer.create(<Todo todo={fakeTodo} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
  // test("renders", () => {
  // const { container, getByText } = render(<Todo todo={fakeTodo} />);
  // console.log(getByText(/edit/i));
  // });
});

const fakeTodo2 = {
  id: "1",
  title: "wrong typed todo",
  complete: "false"
};

console.error = jest.fn();

describe("<Todo /> with react-testing-library", () => {
  test("todo should update", () => {
    const { debug } = render(<Todo todo={fakeTodo2} />);
    // debug();
    expect(console.error).toBeCalled();
  });
});
