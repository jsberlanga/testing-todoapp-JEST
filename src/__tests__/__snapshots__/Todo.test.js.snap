// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> changes class when complete 1`] = `
<div
  className="todo"
>
  <div
    className="todo--option"
  >
    <button
      onClick={[Function]}
    >
      edit
    </button>
    <p
      className="todo__complete"
      onClick={[Function]}
    >
      fake todo
    </p>
    <button
      onClick={[Function]}
    >
      x
    </button>
  </div>
</div>
`;

exports[`<Todo /> renders todo 1`] = `
<div>
  <div
    class="todo"
  >
    <div
      class="todo--option"
    >
      <button>
        edit
      </button>
      <p
        class="todo__complete"
      >
        fake todo
      </p>
      <button>
        x
      </button>
    </div>
  </div>
</div>
`;
